txnkv/transaction/txn.go

```diff
@@ -988,24 +994,26 @@ func (txn *KVTxn) lockKeys(ctx context.Context, lockCtx *tikv.LockCtx, fn func()
 				return err
 			}
 
+			filteredAggressiveLockedKeysCount = len(allKeys) - len(keys)
+			metrics.AggressiveLockedKeysDerived.Add(float64(filteredAggressiveLockedKeysCount))
+			metrics.AggressiveLockedKeysNew.Add(float64(len(keys)))
+
 			if len(keys) == 0 {
+				if lockCtx.Stats != nil {
+					lockCtx.Stats.AggressiveLockDerivedCount += filteredAggressiveLockedKeysCount
```

Comment:
What about moving this out of `if len(keys) == 0`? Then, we make all metrics update and stats update together.

Reply:
Does it matter if `Stats` counts failed ones?

Reply:
Ah, seems better not. I see TiDB merges the stats no matter whether there's an error.